### ISSUE TITLE
[RFC] Clean SGLang Runtime Code from vime

### ISSUE THREAD

**Issue by gcanlin** [vllm-project/vime#7]
# RFC: Clean SGLang Runtime Code from vime

## Summary

vime was originally forked from slime, whose rollout path historically used SGLang as the default inference backend. Now that the vLLM adaptation has landed, vime's runtime target should converge on vLLM and stop maintaining SGLang compatibility.

This RFC proposes cleaning the SGLang runtime code from vime's slime-derived codebase. The cleanup rule is not "delete anything that mentions SGLang." The correct order is:

1. Prefer an equivalent capability in vLLM, vllm-router, or vLLM RLHF weight transfer.
2. If the dependency is generic infrastructure rather than SGLang semantics, reimplement it as a vime-owned utility.
3. Delete only code that is truly SGLang-only: SGLang parameters, protocols, server lifecycle, external SGLang compatibility, and legacy debug entry points.
4. Keep SGLang Docker image build and patch assets, but mark them as docker-only / legacy build assets. They must not imply that vime runtime supports SGLang.

The target state is: vime runtime no longer imports `sglang`, `sglang_router`, or `sglang.srt.*`; the default rollout path is vLLM-only; SGLang is no longer a user-selectable backend.

## Goals

1. Remove the SGLang runtime compatibility surface: `--rollout-backend=sglang`, SGLang engine actors, SGLang rollout functions, and external SGLang server/router support.
2. Migrate valuable capabilities to vLLM equivalents: engine lifecycle, router integration, rollout HTTP APIs, sleep/wake, prefix cache reset, weight update, routed experts replay, and related flows.
3. Move generic utilities into vime-owned code: tensor buckets, multiprocessing serializers, torch reduction patches, quantization helpers, HTTP/router helpers, and deployment config dataclasses.
4. Remove the `--sglang-*` argument system. Preserved capabilities should be renamed to `--vllm-*` or engine-agnostic `--rollout-*` / `--router-*` options.
5. Remove SGLang-specific tests, examples, CI suites, and current user-facing docs. Valuable coverage should be migrated to vLLM tests.
6. Keep SGLang build assets in the Dockerfile design, while clearly separating them from runtime support.

## Non-Goals

1. Do not keep SGLang as a backend, provide fallback behavior, or add old-argument shims.
2. Do not wrap SGLang APIs in a new abstraction and keep using them. APIs such as `ServerArgs`, `launch_server`, `sglang_router`, and `sglang.srt.model_loader` must leave the runtime import graph.
3. Do not blindly delete business capabilities. Capabilities such as logprobs, weight sync, offload, routing replay, multi-model routing, and tool-call parsing must first be classified as vLLM equivalents or vime-owned reimplementations.
4. Do not redesign Docker images in this RFC. This RFC only defines the retention and isolation boundary for Docker assets.

## Current State

vime already contains part of the vLLM adaptation:

| Existing vLLM surface | Current file |
| --- | --- |
| vLLM engine actor | `slime/backends/vllm_utils/vllm_engine.py` |
| vLLM rollout function | `slime/rollout/vllm_rollout.py` |
| vLLM router integration | `slime/ray/rollout.py`, `slime/utils/http_utils.py`, `requirements.txt` with `vllm-router` |
| vLLM weight transfer | `slime/backends/megatron_utils/update_weight/update_weight_from_distributed.py` with `NcclBridge` / packed transfer |
| vLLM CLI entry | `slime/utils/arguments.py` with `--rollout-backend`, `--vllm-*` |

SGLang still remains in these areas:

| Area | Examples | Initial disposition |
| --- | --- | --- |
| Backend selection | `--rollout-backend choices=["sglang", "vllm"]` | Remove the SGLang choice; make vLLM the fixed default |
| Engine actor | `slime/backends/sglang_utils/sglang_engine.py` | Replace with `VLLMEngine`, then delete |
| SGLang `ServerArgs` passthrough | `slime/backends/sglang_utils/arguments.py` | Remove; map required capabilities to vLLM options |
| SGLang deployment config | `slime/backends/sglang_utils/sglang_config.py`, `--sglang-config` | Migrate to vLLM/generic config, or delete unsupported parts |
| Rollout implementation | `slime/rollout/sglang_rollout.py` | Move generic sample/state logic to vime-owned modules; use `vllm_rollout.py` |
| Router | `sglang_router`, `--sglang-router-ip/port` | Migrate to `vllm-router` or a vime router; rename args |
| Weight sync helper | `slime/backends/megatron_utils/sglang.py` | Classify per dependency: vLLM equivalent or vime reimplementation |
| Quantization helpers | SGLang fp8/mxfp8 helper imports | Prefer vLLM / torch / local implementations; otherwise vime-owned |
| Tests and examples | `tests/e2e/*sglang*`, `SGLANG_ARGS` in scripts | Migrate valuable scenarios to vLLM; delete SGLang-only cases |
| Docs | `sglang-config.md`, quick start, blogs, README | Update current user docs to vLLM-only; archive historical docs if needed |
| Docker | `docker/**/sglang.patch`, SGLang base images | Keep as docker-only / legacy assets |

## Cleanup Principles

Every SGLang dependency must be classified in this order. Do not jump directly to deletion.

### 1. Prefer vLLM Equivalent

If vLLM already provides an equivalent capability, migrate to it:

| SGLang dependency | Prefer |
| --- | --- |
| `SGLangEngine` actor | `VLLMEngine` |
| `sglang_router` | `vllm-router` or a vime router |
| SGLang `/generate` | vLLM OpenAI `/v1/completions` or vLLM native generate endpoint |
| `release_memory_occupation` / `resume_memory_occupation` | vLLM `/sleep` / `/wake_up` |
| SGLang weight update group | vLLM RLHF weight transfer / `NCCLWeightTransferEngine` / current `NcclBridge` |
| SGLang prefix/radix cache flush | vLLM prefix cache reset |
| Routed experts replay | vLLM routed experts output if available; otherwise record a vLLM feature gap |
| External SGLang server | External vLLM server |

### 2. Reimplement Generic Utilities in vime

If the code only borrows a generic utility from SGLang, migrate it to a vime-owned implementation instead of continuing to import SGLang:

| Current SGLang import | Proposed vime-owned replacement |
| --- | --- |
| `MultiprocessingSerializer` | `slime/utils/serialization.py` or a local weight-sync serializer |
| `FlattenedTensorBucket` | `slime/backends/megatron_utils/update_weight/tensor_bucket.py` |
| `monkey_patch_torch_reductions` | vime-owned multiprocessing CUDA reducer patch, or removal after validating the PyTorch-native path |
| fp8/mxfp8 helper | vLLM/torch equivalent; if none exists, copy the minimal required logic with source notes and tests |
| SGLang tool-call parser | OpenAI-compatible parser or vime-owned parser |
| Config dataclass | Engine-agnostic deployment config or vLLM config |

After migration, these files must not keep `sglang` naming. Filenames, class names, metric labels, and logs should use vime/generic/vLLM terminology.

### 3. Delete SGLang-Only Compatibility

The following should be deleted:

1. `--rollout-backend=sglang`.
2. Automatic `--sglang-*` passthrough into `ServerArgs`.
3. `sglang_parse_args` and `ServerArgs.add_cli_args` dependency.
4. External SGLang server / router modes.
5. SGLang worker types with no vLLM-equivalent semantics that vime currently supports, such as paths only tailored for SGLang PD/encoder/placeholder flows.
6. SGLang version compatibility branches.
7. SGLang-only debug tools and examples.

### 4. Keep Docker Assets Isolated

SGLang content under Docker may remain, but it must satisfy:

1. It lives only under `docker/**`.
2. README text clearly states it is a legacy/docker-only build asset.
3. Runtime packages, tests, examples, and current user docs do not depend on these patches.
4. The default runtime image does not require `import sglang` to succeed.

### Runtime Package

The runtime package should no longer contain SGLang-specific modules:

```text
slime/backends/sglang_utils/             removed
slime/rollout/sglang_rollout.py          removed
slime/backends/megatron_utils/sglang.py  removed
```

The default rollout entry point is fixed to vLLM:

```text
slime.rollout.vllm_rollout.generate_rollout
```

If `sglang_rollout.py` still contains generic `GenerateState`, sample handling, or logprob post-processing logic, move that logic into `slime/rollout/common_*` or `slime/rollout/generate_utils/*` before deleting the old file. Do not keep a `sglang_rollout.py` shim.

### CLI and Config

Remove all SGLang-facing CLI:

| Old | New / disposition |
| --- | --- |
| `--rollout-backend sglang` | deleted; backend fixed to vLLM |
| `--sglang-router-ip` | `--router-ip` |
| `--sglang-router-port` | `--router-port` |
| `--sglang-router-request-timeout-secs` | `--router-request-timeout-secs` |
| `--sglang-config` | `--vllm-config` or generic `--rollout-config` if semantics are preserved |
| `--sglang-mem-fraction-static` | `--vllm-gpu-memory-utilization` |
| `--sglang-context-length` | `--rollout-max-context-len` / vLLM `--max-model-len` |
| `--sglang-dp-size`, `--sglang-ep-size`, `--sglang-pp-size` | vLLM parallel config if supported; otherwise unsupported |

Do not provide old-argument compatibility. Old commands should fail.

### Router

Router cleanup targets:

1. Runtime no longer imports `sglang_router`.
2. `slime/ray/rollout.py` starts only `vllm-router` or a vime router.
3. Argument names move from `sglang_router_*` to `router_*`.
4. Worker register/list/remove, consistent hashing, and session routing should migrate to `vllm-router` APIs or a vime-owned protocol if still needed.
5. If a `sglang_router` behavior is missing from `vllm-router`, record it as a vLLM router feature gap. Do not keep an SGLang router fallback.

### Weight Update

Weight sync should be handled by category:

| Category | Action |
| --- | --- |
| Existing vLLM equivalent | Use `NCCLWeightTransferEngine`, packed transfer, and current `NcclBridge` |
| Generic trainer-side tools | Implement serializer, tensor bucket, and CUDA IPC/reducer support in vime |
| SGLang model loader / parameter mapper | Do not keep; map to the vLLM weight update contract or delete unsupported paths |

`slime/backends/megatron_utils/sglang.py` should be dismantled:

1. `MultiprocessingSerializer` -> local vime serializer.
2. `FlattenedTensorBucket` -> local vime tensor bucket.
3. fp8/mxfp8 quant helper -> prefer vLLM/torch equivalent; otherwise implement the minimal local logic.
4. `monkey_patch_torch_reductions` -> verify whether it is still required; if yes, implement a local vime patch.

### Rollout

`slime/rollout/vllm_rollout.py` is the target implementation. During cleanup, compare it with `sglang_rollout.py` and confirm that these capabilities are covered:

1. token-id inputs and partial continuation.
2. prompt / completion logprobs.
3. multimodal render + generate.
4. stop / max tokens / sampling params.
5. abort / pause / resume.
6. routed experts replay.
7. OpenAI-compatible output normalization.

Missing capabilities should be added to `vllm_rollout.py` or generic rollout utilities. Do not keep SGLang rollout as a fallback.

### Phase 0: Inventory and Classification

1. Generate the non-docker SGLang dependency list:

```bash
rg -n "sglang|SGLang|SGLANG|sglang_router|sglang\\.srt" \
  slime tests examples scripts docs .github requirements.txt setup.py \
  --glob '!docker/**'
```

2. Classify each hit as:

```text
VLLM_EQUIVALENT
VIME_REIMPLEMENT
DELETE_SGLANG_ONLY
DOCKER_ONLY
ARCHIVE_DOC
```

3. Forbid new non-docker SGLang runtime imports.

### Phase 1: Make vLLM the Only Runtime Backend

1. Remove the `--rollout-backend=sglang` choice and fix the backend to vLLM.
2. Remove `sglang-router` from `requirements.txt`.
3. Remove `sglang_parse_args`, `sglang_validate_args`, and SGLang-router `RouterArgs` imports from `slime/utils/arguments.py`.
4. Rename `args.sglang_router_ip/port` to `args.router_ip/port`.
5. Remove the SGLang/vLLM branching in `slime/ray/rollout.py`; keep the vLLM path.

### Phase 2: Replace Engine and Router

1. Use `slime/backends/vllm_utils/vllm_engine.py` as the only engine actor.
2. Delete `slime/backends/sglang_utils/sglang_engine.py`.
3. Keep only `vllm-router` or a vime router in the router launcher in `slime/utils/http_utils.py`.
4. Align health checks, worker register/list/remove, and PD routing with vLLM router behavior.
5. Decide how to migrate `sglang_config.py`: keep and rename semantics that can be expressed as vLLM/generic deployment config; delete SGLang-only fields.

### Phase 3: Replace Shared Helpers

1. Dismantle `slime/backends/megatron_utils/sglang.py`.
2. Add vime-owned serializer / tensor bucket / torch reduction patch implementations, with unit tests.
3. Prefer vLLM/torch equivalents for quantization helpers. If none exists, copy the minimal required logic into vime and test it.
4. Migrate tool-call parsers, debug comparators, and send helpers to OpenAI/vLLM or vime-owned implementations if they still have business value.

### Phase 4: Finish Rollout and Weight Sync Parity

1. Compare `sglang_rollout.py` with `vllm_rollout.py` and close behavior gaps.
2. Compare SGLang weight update RPCs with vLLM native / packed / reload paths and confirm current training scenarios are covered.
3. If vLLM does not support a capability, record a feature gap or implement it in vime. Do not fall back to SGLang.
4. Delete `slime/rollout/sglang_rollout.py`.

### Phase 5: Tests, Examples, Docs, CI

1. For `tests/e2e/*sglang*` and `tests/utils/test_sglang_config.py`, migrate valuable coverage to vLLM and delete SGLang-only cases.
2. Remove or convert CI entries such as `e2e-test-sglang-config`, `run-ci-sglang-config`, and SGLang checkout/install steps.
3. Convert `pkill -9 sglang`, `SGLANG_ARGS`, and `--sglang-*` in scripts/examples to vLLM; delete SGLang-only examples.
4. Update current user docs to vLLM-only. Historical blogs may remain as archives but must not appear in quick start or current support paths.
5. Mark SGLang Docker patches as legacy/docker-only in the Docker README.

## Acceptance Criteria

Code acceptance:

1. The non-docker runtime import graph has no `sglang`, `sglang_router`, or `sglang.srt.*`.
2. `python -c "import slime"` succeeds in an environment without SGLang installed.
3. `requirements.txt` does not contain `sglang-router`.
4. CLI no longer exposes `--rollout-backend=sglang` or `--sglang-*`.
5. Default training/rollout starts only vLLM.

Search acceptance:

```bash
rg -n "from sglang|import sglang|sglang_router|sglang\\.srt|--sglang-|rollout-backend.*sglang" \
  slime tests examples scripts docs .github requirements.txt setup.py \
  --glob '!docker/**'
```

Allowed hits:

```text
docker/**
docs/rfcs/0001-clean-sglang*.md
archived historical docs, if explicitly marked archive-only
```

Functional acceptance:

1. fast tests pass.
2. vLLM rollout smoke test passes.
3. vLLM weight update smoke test passes, covering at least native or packed mode.
4. colocate offload/onload smoke test passes.
5. router worker register/list/remove smoke test passes.
6. Docs no longer claim vime supports SGLang runtime.

## Risks and Mitigations

| Risk | Impact | Mitigation |
| --- | --- | --- |
| Generic helpers are deleted by mistake | Training or weight-sync regressions | Classify each dependency first; prefer vLLM equivalents and vime reimplementation before deletion |
| vLLM equivalent behavior is not identical | Rollout/logprob/weight-sync drift | Maintain a parity checklist and smoke tests |
| vLLM router lacks an SGLang router behavior | Session/routing/PD scenarios regress | Record a feature gap; implement in vime router if needed; do not fall back to SGLang |
| Quantization helper has no ready equivalent | fp8/mxfp8 path fails | Copy the minimal required logic into vime with source notes and tests |
| Docker still mentions SGLang and causes confusion | Support boundary is unclear | Mark Docker assets as legacy/docker-only; runtime acceptance excludes docker |

## Rollout Strategy

Split the work by dependency direction so each step remains verifiable:

1. PR 1: Add classification docs and a CI grep guard forbidding new SGLang runtime imports.
2. PR 2: Fix vLLM as the backend; clean up CLI and router naming.
3. PR 3: Replace `megatron_utils/sglang.py` with vime-owned helpers.
4. PR 4: Close vLLM rollout / weight sync parity gaps; delete SGLang engine and rollout.
5. PR 5: Migrate tests/examples/docs/CI and mark Docker assets as legacy.

## Open Questions

1. Does current vLLM weight update cover all production paths: tensor update, distributed update, packed transfer, MoE experts, and reload fallback?
2. Which `--sglang-config` semantics, such as multi-model, PD, and heterogeneous server groups, should become `--vllm-config` / `--rollout-config`?
3. Is the vLLM returned format for R3 / routed experts replay stable enough for the training-side consumer?
4. Which historical docs/blogs should remain as archives, and which should be removed from current docs?
